handler/preprocess.sh

```diff
@@ -35,6 +36,48 @@ echo "finding dicom directories"
 ./find_dicomdir.py $root > $root/dcm2niix.list
 cat $root/dcm2niix.list
 
+# There can be overlap between the dicom directories and the pet directories
+# we want to use dcm2niix4pet for the pet directories, and dcm2niix for the dicom directories
+# but if and only if the conversion library PET2BIDS is installed
+PET2BIDS_INSTALLED=`which dcm2niix4pet > /dev/null && echo $?`
+
+if [ $PET2BIDS_INSTALLED -eq 0 ]; then
+
+    echo "PET2BIDS is installed, using dcm2niix4pet for PET directories found in root dir ${root}"
+    function rundcm2niix4pet {
+        #note.. this function runs inside $root (by --wd $root)
+
+        path=$1
+
+        echo "----------------------- $path ------------------------"
+        timeout 3600 dcm2niix4pet $path
+
+        #all good
+        echo $path >> pet2bids.done
```

Comment:
This is problematic when running on Linux, since it will use `parallel` to execute the function. Parallel processes may try to write at the same time, messing up the content of the file.

One possibility is to just echo to stdout and use `--group` in the parallel command. Then the output of the parallel program can be saved to `pet2bids.done`. This will guarantee that 1. the paths will be correct on the file, one per line, and 2. it will only write on exit, which makes more sense.

Reply:
So is this handler/preprocess.sh script going to be called on the same directory in parallel? Or are you referring to this?

`cat $root/dcm2niix.list | parallel --linebuffer --wd $root -j 6 d2n {} 2>> $root/dcm2niix_output`

It's my understanding that `$root` is a unique dir corresponding to and upload/dump of dicom images, is that not correct?

Reply:
Yes, I am referring to the `parallel` command and your understanding is correct, `$root` is a unique directory of file upload.

My point is about the mechanism used to store which files were processed, that is writing the file path into `pet2bids.done`

E.g. if I am running in parallel two files `abracadabra.dcm` and `alakazam.dcm`. One would expect that the final state of the `pet2bids.done` is:
```
alakazam.dcm
abracadabra.dcm
```

However, given the race condition nature of parallelism, the file can become:
```
alabrackazaadabra.dcm
m.dcm
```

Reply:
Okay that sounds good. But isn't `--group` the default for parallel in the above instance? 

`-j` is 6 jobs max
`--linebuffer` only outputs full lines
`--wd` is working dir

So, isn't:

`cat $root/dcm2niix.list | parallel --linebuffer --wd $root -j 6 d2n {} 2>> $root/dcm2niix_output`

equivalent to 

`cat $root/dcm2niix.list | parallel --linebuffer --group --wd $root -j 6 d2n {} 2>> $root/dcm2niix_output` ?

Or am I missing something here? I haven't used gnu parallel at all.